contrib/epee/include/net/levin_protocol_handler_async.h

```diff
@@ -769,36 +769,32 @@ void async_protocol_handler_config<t_connection_context>::del_connection(async_p
 template<class t_connection_context>
 void async_protocol_handler_config<t_connection_context>::delete_connections(size_t count, bool incoming)
 {
-  std::vector <boost::uuids::uuid> connections;
+  std::vector<typename connections_map::mapped_type> connections;
+
+  auto scope_exit_handler = misc_utils::create_scope_leave_handler([&connections]{
+    for (auto &aph: connections)
+      aph->finish_outer_call();
+  });
+
   CRITICAL_REGION_BEGIN(m_connects_lock);
   for (auto& c: m_connects)
   {
     if (c.second->m_connection_context.m_is_income == incoming)
-      connections.push_back(c.first);
+      if (c.second->start_outer_call())
```

Comment:
Do the test cases below trigger something? The object has to wait in the destructor for `m_connects_lock`, which is somewhat gross, but it cannot be fully cleaned up. It looks safer to add this, but I'm not sure that its strictly necessary (probably a red herring for problems you've observed).

The `shared_ptr`/`weak_ptr` patch also does this, but its mandatory to get the real pointer from the `weak_ptr`.